### ISSUE TITLE
How reliable the output is?

### ISSUE THREAD

**Issue by HelloWorldLTY** [gerstung-lab/Delphi#25]
Hi, thanks for your interesting work. I am reproducing the example results, but find a interesting result:

72.7: D50 Iron deficiency anaemia
72.9: No event
72.9: E78 Disorders of lipoprotein metabolism and other lipidaemias

In 72.9, the 1st prediction is no event, but the next token is still in the same time scale, and it happens a new event. How to interpret it? Does this mean we can predict multiple events for a single time point? If so, how reliable the output is?

**Comment by yozhikoff** [comment]
Hi!

If I understand correctly, what seems confusing is that there is "no event" and "some event" at the same time. 
I'll start with this question. 

> How to interpret it? If so, how reliable the output is?

The seeming co-occurrence of two tokens is just a coincidence (see below for more details). It should not be interpreted as something special. Additionally, "no event" tokens are inserted into the training data at random positions (see the paper for the motivation behind them), so they do imply that there will be no records during that period.

And regarding how reliable the output is - it's quite good in terms of determining risks, but the samples are always stochastic. Health itself is stochastic, so one shouldn't expect that it's possible to predict *exact* dates of health events. (aka D50 Iron deficiency anaemia *exactly* at the age of 72.7)

> Does this mean we can predict multiple events for a single time point?

Not with the current formulation of the model. The sampling process occurs by sampling the waiting time until the next token from an exponential distribution with a rate parameter associated with the predicted logits. For the exponential distribution, samples are always strictly above zero, so two tokens cannot co-occur exactly at the same time. What likely happened in your example is that the difference between the two events is so small that rounding makes it seem the same. Perhaps with less rounding, it would be something like (extreme example):

72.7000: D50 Iron deficiency anaemia
72.9000: No event
72.9005: E78 Disorders of lipoprotein metabolism and other lipidaemias
 
The sampling procedure is described better in the beginning of the [sampling notebook](https://github.com/gerstung-lab/Delphi/blob/main/sampling_trajectories.ipynb)


I hope this helps! Please don't hesitate to ask more if the above doesn't fully address your questions.

**Comment by HelloWorldLTY** [comment]
Thanks, this is very informative!
